### PR TITLE
feat(eval): Add semantic similarity / entailment evaluators, QnA gen

### PR DESCRIPTION
**Pull request by jon-chuang**
# Description
Eval e2e

Notebook: https://github.com/jerryjliu/llama_index/blob/369ebcc61e439d220b69ba5b6935f48ceace2240/docs/examples/evaluation/QnA-Gen-Answer-Similarity.ipynb

The notebook should render differently in sphinx
image

## Type of Change

- [X] New feature (non-breaking change which adds functionality)
- [ ] This change requires a documentation update

# How Has This Been Tested?
- [X] Added new notebook (that tests end-to-end)

# Suggested Checklist:

- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

File: llama_index/evaluation/dataset_generation.py [hunk 1]
```diff
@@ -131,21 +138,39 @@ def _node_question_generator(
...
 
         if num is not None:
-            questions = questions[:num]
-        return questions
+            output_strings = output_strings[:num]
+        return output_strings
 
     def generate_questions_from_nodes(self, num: Optional[int] = None) -> List[str]:
         """Generates questions for each document."""
-        return self._node_question_generator(self.nodes, num)
```
Comment (logan-markewich) [inline llama_index/evaluation/dataset_generation.py:151]: maybe we can keep the old name, but use a boolean toggle like `generate_questions_from_nodes(include_answers=True)`

File: llama_index/evaluation/dataset_generation.py [hunk 2]
```diff
@@ -131,21 +138,39 @@ def _node_question_generator(
...
+
+    def generate_qna_from_nodes(
+        self, num: Optional[int] = None
+    ) -> List[Tuple[str, str]]:
+        """Generates questions for each document."""
+        output_strings = self._node_qna_generator(self.nodes, num, qna_mode=True)
+        # Validate that questions start with Q: and answers start with A:
+        for i in range(0, len(output_strings), 2):
+            if not output_strings[i].startswith("Q:"):
+                raise ValueError(f"Question {output_strings[i]} does not start with Q:")
```
Comment (logan-markewich) [inline llama_index/evaluation/dataset_generation.py:168]: Might be better to print a warning and skip the question

File: llama_index/evaluation/scoring/semantic_answer_similarity.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+                "please `pip install sentence-transformers`",
+            )
+
+        self._model = CrossEncoder(model, max_length=512)
+
+    def evaluate(self, ground_truth: str, predicted: str) -> float:
+        """
+        Evaluates the semantic similarity between a ground truth answer and
+        a predicted answer.
+        """
```
Comment (logan-markewich) [inline llama_index/evaluation/scoring/semantic_answer_similarity.py:44]: This has the same doctoring as `SemanticAnswerSimilarity`  -- what's the actual difference here? Why would I use one vs. the other?

I see that this is an NLI model (i.e. predicts entails, contradicts, neutral), but might be good to better convey to a user how this differs from normal similarity 

Some possible ideas:
- `SemanticAnswerAgreement`
- I thought I had more idea but tbh I'm drawing blanks haha

**Comment by jon-chuang** [comment]
Generally, while this seems useful, I wonder if it might be possible to use off the shelf solutions (e.g. `huggingface.evaluate`)

**Comment by jerryjliu** [comment]
@jon-chuang  i'm still down to add these capabilities to expand our evaluation presence. would be good to address @logan-markewich 's comments though

**Comment by logan-markewich** [comment]
Just a small bump, might be nice to get this wrapped up 💪🏻

**Comment by logan-markewich** [comment]
Actually, I'll get this merged in the next little bit. I think this could be immensely useful for evaluating queries 🙏🏻

**Comment by Disiok** [comment]
Closing since it's outdated.

**Pull request closed without merging**